src/System.Buffers/tests/ArrayPool/CollectionTests.cs

```diff
@@ -0,0 +1,193 @@
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Xunit;
+
+namespace System.Buffers.ArrayPool.Tests
+{
+    [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+    public class CollectionTests : ArrayPoolTest
+    {
+        [OuterLoop("This is a long running test (over 2 minutes)")]
+        [Theory,
+            InlineData(true),
+            InlineData(false)]
+        public void BuffersAreCollectedWhenStale(bool trim)
+        {
+            RemoteInvokeWithTrimming((trimString) =>
+            {
+                // Check that our environment is as we expect
+                Assert.Equal(trimString, Environment.GetEnvironmentVariable(TrimSwitchName));
+
+                const int BufferCount = 8;
+                const int BufferSize = 1025;
+
+                // Get the pool and check our trim setting
+                var pool = ArrayPool<int>.Shared;
+                bool parsedTrim = ValidateTrimState(pool, trimString);
+
+                List<int[]> rentedBuffers = new List<int[]>();
+
+                // Rent and return a set of buffers
+                for (int i = 0; i < BufferCount; i++)
+                {
+                    rentedBuffers.Add(pool.Rent(BufferSize));
+                }
+                for (int i = 0; i < BufferCount; i++)
+                {
+                    pool.Return(rentedBuffers[i]);
+                }
+
+                // Rent what we returned and ensure they are the same
+                for (int i = 0; i < BufferCount; i++)
+                {
+                    var buffer = pool.Rent(BufferSize);
+                    Assert.Contains(rentedBuffers, item => ReferenceEquals(item, buffer));
+                }
+                for (int i = 0; i < BufferCount; i++)
+                {
+                    pool.Return(rentedBuffers[i]);
+                }
+
+                // Now wait a little over a minute and force a GC to get some buffers returned
+                Console.WriteLine("Waiting a minute for buffers to go stale...");
+                Thread.Sleep(61 * 1000);
+                GC.Collect(2);
+                GC.WaitForPendingFinalizers();
+                bool foundNewBuffer = false;
+                for (int i = 0; i < BufferCount; i++)
+                {
+                    var buffer = pool.Rent(BufferSize);
+                    if (!rentedBuffers.Any(item => ReferenceEquals(item, buffer)))
+                    {
+                        foundNewBuffer = true;
+                    }
+                }
+
+                // Should only have found a new buffer if we're trimming
+                Assert.Equal(parsedTrim, foundNewBuffer);
+                return SuccessExitCode;
+            }, trim, 3 * 60 * 1000); // This test has to wait for the buffers to go stale (give it three minutes)
+        }
+
+        [OuterLoop("This test is risky as it allocates a ton of memory (can cause other tests to fail)")]
```

Comment:
> can cause other tests to fail

Then we shouldn't have it even in outerloop.

Reply:
The tests that are stressing memory management in different ways (and take over the machine as side-effect) may be better suited for CoreCLR repo. I believe we have other GC tests in this category.

Reply:
Would this create issues in a stress environment too? The problem is running test assemblies in parallel.

Reply:
> Would this create issues in a stress environment too? The problem is running test assemblies in parallel.

We don't have a stress environment for these.  Enabling tests with the suggested attribute currently requires a dev to do it explicitly.  If/when we have better CI support for it, we'll need to address the fact that most such stress tests behave as if they own the box.

Reply:
> My top-level point here is I don't want such tests running as part of outerloop jobs.

Yes, I have no problem with that. Just don't want to kick the ball forward to another currently running automated system. :)